testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java

```diff
@@ -73,16 +75,35 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) {
                             try {
                                 if (msg instanceof ByteBuf) {
                                     aggregator.writeBytes((ByteBuf) msg);
-                                    if (aggregator.readableBytes() == EXPECTED_BYTES) {
-                                        if (clientReceived.compareAndSet(null, aggregator)) {
-                                            latch.countDown();
-                                        }
-                                    }
                                 }
                             } finally {
                                 ReferenceCountUtil.release(msg);
                             }
                         }
+
+                        @Override
+                        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                            // IOException is fine as it will also close the channel and may just be a connection reset.
+                            if (!(cause instanceof IOException)) {
+                                clientReceived.set(cause);
+                                latch.countDown();
+                            }
+                        }
+
+                        @Override
+                        public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+                            if (clientReceived.compareAndSet(null, aggregator)) {
```

Comment:
do we need to worry about channel inactive being called multiple times?

Reply:
@Scottmitch no really but I wanted to ensure we not override the exception that we may have set in `exceptionCaught(...)` before.